src/main/java/org/truffleruby/core/module/ModuleNodes.java

```diff
@@ -1864,6 +1871,79 @@ public DynamicObject setMethodVisibility(DynamicObject module, Object name,
 
     }
 
+    @CoreMethod(names = "refine", needsBlock = true, required = 1, visibility = Visibility.PRIVATE)
+    public abstract static class RefineNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private YieldNode yield = new YieldNode(DeclarationContext.CLASS_EVAL);
+        @Child private CallDispatchHeadNode newModuleNode = CallDispatchHeadNode.createOnSelf();
+
+
+        @Specialization
+        public DynamicObject refine(VirtualFrame frame, DynamicObject self, DynamicObject classToRefine, NotProvided block) {
+            throw new RaiseException(coreExceptions().argumentError("no block given", this));
+        }
+
+        @Specialization
+        public DynamicObject refine(VirtualFrame frame, DynamicObject self, DynamicObject classToRefine, DynamicObject block) {
+            // TODO BJF add block is not a proc error
+
+            if (!RubyGuards.isRubyClass(classToRefine)) {
+                throw new RaiseException(coreExceptions().typeErrorWrongArgumentType(classToRefine, "Class", this));
+            }
+
+//            InternalMethod method = getContext().getCallStack().getCallingMethodIgnoringSend();
+//            LexicalScope lexicalScope = method == null ? null : method.getSharedMethodInfo().getLexicalScope();
+
+            ConcurrentMap<DynamicObject, DynamicObject> refinements = Layouts.MODULE.getFields(self).getRefinements();
+            ConcurrentMap<DynamicObject, DynamicObject> activatedRefinements = Layouts.MODULE.getFields(self).getRefinements();
```

Comment:
Should this be `getActivatedRefinements()`?

Reply:
Yes, fixed.